src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -937,6 +987,21 @@ private bool TryCreateSegmentForNavigationSource(string identifier, string paren
             {
                 this.parsedSegments.Add(segment);
                 this.TryBindKeyFromParentheses(parenthesisExpression);
+
+                if (tryBindingEscapeFunction && !this.TryBindEscapeFunction())
+                {

```

Comment:
So the logic is, for each segment that could be followed by an escape function, if the identifier ends in ':', try binding the identifier up to the colon as that segment type, and see if the you can bind a function to it. If not, revert binding the segment and continue.  Right?

Reply:
Yes, you understanding is correct regarding what I am trying to do here. However, the segment may successfully bind as type which may not have escape function bound but the segment can then be bound to a type which may have an escape function. Therefore, it is incorrect to say that it will only bind to the key segment but yes, it is very likely that it will bind to the key segment. 

Even if I take that assumption, the code will be simplified but not to a great extent.

---
In reply to: [366156392](https://github.com/OData/odata.net/pull/1621#discussion_r366156392) [](ancestors = 366156392)